File: sort-lmfdb.py
#!/usr/bin/env python3
"""
Sort LMFDB mathematical objects into 71 Monster shards by prime resonance
"""

import json
import os
from pathlib import Path
from collections import defaultdict

MONSTER_PRIMES = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 41, 47, 59, 71]

def calculate_prime_resonance(obj):
    """Calculate which prime(s) an object resonates with"""
    resonances = []
    
    # Extract numerical properties
    numbers = extract_numbers(obj)
    
    for prime in MONSTER_PRIMES:
        score = 0
        for num in numbers:
            if num % prime == 0:
                score += 1
        
        if score > 0:
            resonances.append((prime, score / len(numbers) if numbers else 0))
    
    return resonances

def extract_numbers(obj):
    """Extract all numerical values from an object"""
    numbers = []
    
    if isinstance(obj, dict):
        for key, value in obj.items():
            # Check for conductor, level, degree, rank, etc.
            if isinstance(value, int) and value > 0:
                numbers.append(value)
            elif isinstance(value, list):
                numbers.extend([v for v in value if isinstance(v, int) and v > 0])
            elif isinstance(value, dict):
                numbers.extend(extract_numbers(value))
    
    return numbers

def assign_to_shard(obj):
    """Assign object to primary shard based on strongest resonance"""
    resonances = calculate_prime_resonance(obj)
    
    if not resonances:
        return 1  # Default to shard 1
    
    # Sort by resonance score
    resonances.sort(key=lambda x: x[1], reverse=True)
    
    return resonances[0][0]  # Return prime with highest resonance

def process_lmfdb_file(filepath, output_dir):
    """Process a single LMFDB JSON file"""
    print(f"Processing {filepath}...")
    
    with open(filepath, 'r') as f:
        data = json.load(f)
    
    # Initialize shard buckets
    shards = defaultdict(list)
    
    # Sort objects into shards
    if isinstance(data, list):
        for obj in data:
            shard = assign_to_shard(obj)
            shards[shard].append(obj)
    elif isinstance(data, dict):
        for key, obj in data.items():
            shard = assign_to_shard(obj)
            obj['_id'] = key
            shards[shard].append(obj)
    
    # Write to shard files
    filename = Path(filepath).stem
    for shard_num, objects in shards.items():
        shard_dir = output_dir / f"shard-{shard_num:02d}" / "data" / "lmfdb"
        shard_dir.mkdir(parents=True, exist_ok=True)
        
        output_file = shard_dir / f"{filename}.json"
        with open(output_file, 'w') as f:
            json.dump(objects, f, indent=2)
        
        print(f"  Shard {shard_num}: {len(objects)} objects → {output_file}")
    
    return shards

def main():
    lmfdb_dir = Path("/mnt/data1/meta-introspector/data/lmfdb-collected")
    output_dir = Path("/home/mdupont/experiments/monster/monster-shards")
    
    print("🔷 Sorting LMFDB by Prime Resonance")
    print("=" * 50)
    print()
    
    total_stats = defaultdict(int)
    
    # Process each LMFDB file
    for json_file in lmfdb_dir.glob("*.json"):
        shards = process_lmfdb_file(json_file, output_dir)
        
        for shard_num, objects in shards.items():
            total_stats[shard_num] += len(objects)
        
        print()
    
    # Print summary
    print("=" * 50)
    print("Summary:")
    print()
    
    for shard_num in sorted(total_stats.keys()):
        is_prime = shard_num in MONSTER_PRIMES
        marker = "★" if is_prime else " "
        print(f"  Shard {shard_num:2d} {marker}: {total_stats[shard_num]:4d} objects")
    
    print()
    print(f"Total: {sum(total_stats.values())} objects distributed across {len(total_stats)} shards")
    print()
    print("✅ LMFDB sorted by Monster prime resonance!")

if __name__ == "__main__":
    main()

File: ml/llm_analysis/ollama-monster/src/bin/sort-lmfdb.rs
//! Sort LMFDB mathematical objects into 71 Monster shards by prime resonance

use anyhow::Result;
use serde_json::Value;
use std::collections::HashMap;
use std::fs;
use std::path::Path;

const MONSTER_PRIMES: [u32; 15] = [2,3,5,7,11,13,17,19,23,29,31,41,47,59,71];

fn main() -> Result<()> {
    println!("🔷 Sorting LMFDB by Prime Resonance");
    println!("====================================\n");
    
    let lmfdb_dir = Path::new("/mnt/data1/meta-introspector/data/lmfdb-collected");
    let output_dir = Path::new("monster-shards");
    
    let mut total_stats: HashMap<u32, usize> = HashMap::new();
    
    // Process each JSON file
    for entry in fs::read_dir(lmfdb_dir)? {
        let path = entry?.path();
        if path.extension().and_then(|s| s.to_str()) == Some("json") {
            let shards = process_lmfdb_file(&path, output_dir)?;
            
            for (shard, count) in shards {
                *total_stats.entry(shard).or_insert(0) += count;
            }
        }
    }
    
    // Print summary
    println!("\n====================================");
    println!("Summary:\n");
    
    let mut shards: Vec<_> = total_stats.iter().collect();
    shards.sort_by_key(|(k, _)| *k);
    
    for (shard, count) in shards {
        let marker = if MONSTER_PRIMES.contains(shard) { "★" } else { " " };
        println!("  Shard {:2} {}: {:4} objects", shard, marker, count);
    }
    
    let total: usize = total_stats.values().sum();
    println!("\nTotal: {} objects across {} shards", total, total_stats.len());
    println!("\n✅ LMFDB sorted by Monster prime resonance!");
    
    Ok(())
}

fn process_lmfdb_file(path: &Path, output_dir: &Path) -> Result<HashMap<u32, usize>> {
    println!("Processing {:?}...", path.file_name().unwrap());
    
    let content = fs::read_to_string(path)?;
    let data: Value = serde_json::from_str(&content)?;
    
    let mut shards: HashMap<u32, Vec<Value>> = HashMap::new();
    
    // Sort objects into shards
    match data {
        Value::Array(objects) => {
            for obj in objects {
                let shard = assign_to_shard(&obj);
                shards.entry(shard).or_default().push(obj);
            }
        }
        Value::Object(map) => {
            for (key, obj) in map {
                let shard = assign_to_shard(&obj);
                let mut obj_with_id = obj.clone();
                if let Value::Object(ref mut m) = obj_with_id {
                    m.insert("_id".to_string(), Value::String(key));
                }
                shards.entry(shard).or_default().push(obj_with_id);
            }
        }
        _ => {}
    }
    
    // Write to shard files
    let filename = path.file_stem().unwrap().to_str().unwrap();
    let mut counts = HashMap::new();
    
    for (shard_num, objects) in &shards {
        let shard_dir = output_dir
            .join(format!("shard-{:02}", shard_num))
            .join("data")
            .join("lmfdb");
        
        fs::create_dir_all(&shard_dir)?;
        
        let output_file = shard_dir.join(format!("{}.json", filename));
        let json = serde_json::to_string_pretty(objects)?;
        fs::write(&output_file, json)?;
        
        counts.insert(*shard_num, objects.len());
        println!("  Shard {}: {} objects", shard_num, objects.len());
    }
    
    Ok(counts)
}

fn assign_to_shard(obj: &Value) -> u32 {
    let resonances = calculate_prime_resonance(obj);
    
    if resonances.is_empty() {
        return 1; // Default to shard 1
    }
    
    // Return prime with highest resonance
    resonances.iter()
        .max_by(|a, b| a.1.partial_cmp(&b.1).unwrap())
        .map(|(prime, _)| *prime)
        .unwrap_or(1)
}

fn calculate_prime_resonance(obj: &Value) -> Vec<(u32, f64)> {
    let numbers = extract_numbers(obj);
    
    if numbers.is_empty() {
        return vec![];
    }
    
    let mut resonances = Vec::new();
    
    for &prime in &MONSTER_PRIMES {
        let score: usize = numbers.iter()
            .filter(|&&n| n % prime == 0)
            .count();
        
        if score > 0 {
            let rate = score as f64 / numbers.len() as f64;
            resonances.push((prime, rate));
        }
    }
    
    resonances
}

fn extract_numbers(obj: &Value) -> Vec<u32> {
    let mut numbers = Vec::new();
    extract_numbers_recursive(obj, &mut numbers);
    numbers
}

fn extract_numbers_recursive(obj: &Value, numbers: &mut Vec<u32>) {
    match obj {
        Value::Number(n) => {
            if let Some(i) = n.as_u64() {
                if i > 0 && i <= u32::MAX as u64 {
                    numbers.push(i as u32);
                }
            }
        }
        Value::Array(arr) => {
            for item in arr {
                extract_numbers_recursive(item, numbers);
            }
        }
        Value::Object(map) => {
            for (_, value) in map {
                extract_numbers_recursive(value, numbers);
            }
        }
        _ => {}
    }
}
